File: src/arc-shapes/draw-comet-path.js
import { ARCSHAPE, BEHAVIOR } from "../utils/constants";
import { getCircularArcLocus, getSpiralArcLocus } from "../utils/formulas";
import { PathCommand } from "../utils/path-command";
import { drawArcs, radiansToDegree } from "../utils/utils";

/**
 * Draw comet and reverse-comet shape.
 * 
 * @param {number} radius 
 * @param {number} width 
 * @param {number} margin 
 * @param {boolean} morphingShape 
 * @param {string} shape 
 * @param {number} startingDegree 
 * @param {number} endDegree 
 * @returns {string} SVG path
 */
export default function drawCometPath(radius, width, margin, morphingShape, shape, startingDegree, endDegree) {
    const gap = radiansToDegree(width / radius);

    let arc1, arc2;
    if (shape === ARCSHAPE.COMET) {
        arc1 = getSpiralArcLocus(1, radius, width, margin, BEHAVIOR.GROW, startingDegree, endDegree - gap);
        arc2 = getSpiralArcLocus(3, radius, width, margin, BEHAVIOR.GROW, endDegree - gap, startingDegree);
    } else if (shape === ARCSHAPE.REVERSE_COMET) {
        arc1 = getSpiralArcLocus(1, radius, width, margin, BEHAVIOR.SHRINK, startingDegree + gap, endDegree);
        arc2 = getSpiralArcLocus(3, radius, width, margin, BEHAVIOR.SHRINK, endDegree, startingDegree + gap);
    }

    const arc1Start = arc1[0];
    const arc1End = arc1[arc1.length - 1];

    const arc2Start = arc2[0];
    const arc2End = arc2[arc2.length - 1];

    const arc3 = getCircularArcLocus(1, radius, width, margin, morphingShape, startingDegree, endDegree);
    const arc3Start = arc3[0];
    const arc3End = arc3[arc3.length - 1];

    const arc4 = getCircularArcLocus(3, radius, width, margin, morphingShape, endDegree, startingDegree);
    const arc4Start = arc4[0];
    const arc4End = arc4[arc4.length - 1];

    const arc5 = getCircularArcLocus(2, radius, width, margin, morphingShape, startingDegree, endDegree);
    const arc5Start = arc5[0];
    const arc5End = arc5[arc5.length - 1];

    /*
        Draw path pattern:
        without morph           : M -> A1 -> C1 -> C2 -> A2
        without morph (reverse) : M -> A1 ->             A2 -> C3 -> C4
        with morph              : M -> A1 -> C1 -> C2 -> A2 -> C3 -> C4
    */
    let M, A1, C1, C2, A2, C3, C4;

    M = PathCommand.moveTo.M(arc1Start.point.x, arc1Start.point.y);
    A1 = drawArcs(morphingShape, 0, 0, arc1.slice(1));
    A2 = drawArcs(morphingShape, 0, 1, arc2.slice(1));

    if (morphingShape) {
        if (shape === ARCSHAPE.COMET) {
            C1 = PathCommand.cubicBezierCurve.C(arc3End.point.x, arc3End.point.y, arc5End.point.x, arc5End.point.y, arc5End.point.x, arc5End.point.y);

            C2 = PathCommand.cubicBezierCurve.C(arc5End.point.x, arc5End.point.y, arc4Start.point.x, arc4Start.point.y, arc2Start.point.x, arc2Start.point.y);

            C3 = PathCommand.cubicBezierCurve.C(arc2End.point.x, arc2End.point.y, arc2End.point.x, arc2End.point.y, arc2End.point.x, arc2End.point.y);

            C4 = PathCommand.cubicBezierCurve.C(arc2End.point.x, arc2End.point.y, arc1Start.point.x, arc1Start.point.y, arc1Start.point.x, arc1Start.point.y);
        } else if (shape === ARCSHAPE.REVERSE_COMET) {
            C1 = PathCommand.cubicBezierCurve.C(arc1End.point.x, arc1End.point.y, arc1End.point.x, arc1End.point.y, arc1End.point.x, arc1End.point.y);

            C2 = PathCommand.cubicBezierCurve.C(arc1End.point.x, arc1End.point.y, arc2Start.point.x, arc2Start.point.y, arc2Start.point.x, arc2Start.point.y);

            C3 = PathCommand.cubicBezierCurve.C(arc4End.point.x, arc4End.point.y, arc5Start.point.x, arc5Start.point.y, arc5Start.point.x, arc5Start.point.y);

            C4 = PathCommand.cubicBezierCurve.C(arc5Start.point.x, arc5Start.point.y, arc3Start.point.x, arc3Start.point.y, arc1Start.point.x, arc1Start.point.y);
        }

        return `${M} ${A1} ${C1} ${C2} ${A2} ${C3} ${C4}`;
    } else {
        if (shape === ARCSHAPE.COMET) {
            C1 = PathCommand.cubicBezierCurve.C(arc3End.point.x, arc3End.point.y, arc5End.point.x, arc5End.point.y, arc5End.point.x, arc5End.point.y);

            C2 = PathCommand.cubicBezierCurve.C(arc5End.point.x, arc5End.point.y, arc4Start.point.x, arc4Start.point.y, arc2Start.point.x, arc2Start.point.y);

            return `${M} ${A1} ${C1} ${C2} ${A2}`;
        } else if (shape === ARCSHAPE.REVERSE_COMET) {
            C3 = PathCommand.cubicBezierCurve.C(arc4End.point.x, arc4End.point.y, arc5Start.point.x, arc5Start.point.y, arc5Start.point.x, arc5Start.point.y);

            C4 = PathCommand.cubicBezierCurve.C(arc5Start.point.x, arc5Start.point.y, arc3Start.point.x, arc3Start.point.y, arc1Start.point.x, arc1Start.point.y);

            return `${M} ${A1} ${A2} ${C3} ${C4}`;
        }
    }

}

File: wasm/src/arc_shape/draw_comet_path.rs
use crate::util::{
    constant::ArcShape,
    draw_arcs,
    formula::{
        get_circular_arc_locus, get_spiral_arc_locus, Behavior, CircularArcLayer, SpiralArcLayer,
    },
    path_command, radians_to_degree,
};

#[allow(non_snake_case)]
pub fn draw_comet_path(
    radius: f64,
    width: f64,
    margin: f64,
    morphing_shape: bool,
    shape: ArcShape,
    starting_degree: f64,
    end_degree: f64,
) -> String {
    let gap = radians_to_degree(width / radius);

    let (behavior, arc1_degree, arc2_degree) = match shape {
        ArcShape::Comet => {
            let arc1 = (starting_degree, end_degree - gap);
            let arc2 = (end_degree - gap, starting_degree);

            (Behavior::Grow, arc1, arc2)
        }
        ArcShape::ReverseComet => {
            let arc1 = (starting_degree + gap, end_degree);
            let arc2 = (end_degree, starting_degree + gap);

            (Behavior::Shrink, arc1, arc2)
        }
        _ => panic!("Shape not allowed"),
    };

    let arc1 = get_spiral_arc_locus(
        SpiralArcLayer::L1,
        radius,
        width,
        margin,
        behavior,
        arc1_degree.0,
        arc1_degree.1,
    );
    let arc1_start = arc1[0];
    let arc1_end = arc1[arc1.len() - 1];

    let arc2 = get_spiral_arc_locus(
        SpiralArcLayer::L3,
        radius,
        width,
        margin,
        behavior,
        arc2_degree.0,
        arc2_degree.1,
    );
    let arc2_start = arc2[0];
    let arc2_end = arc2[arc2.len() - 1];

    let arc3 = get_circular_arc_locus(
        CircularArcLayer::L1,
        radius,
        width,
        margin,
        morphing_shape,
        starting_degree,
        end_degree,
    );
    let arc3_start = arc3[0];
    let arc3_end = arc3[arc3.len() - 1];

    let arc4 = get_circular_arc_locus(
        CircularArcLayer::L3,
        radius,
        width,
        margin,
        morphing_shape,
        end_degree,
        starting_degree,
    );
    let arc4_start = arc4[0];
    let arc4_end = arc4[arc4.len() - 1];

    let arc5 = get_circular_arc_locus(
        CircularArcLayer::L2,
        radius,
        width,
        margin,
        morphing_shape,
        starting_degree,
        end_degree,
    );
    let arc5_start = arc5[0];
    let arc5_end = arc5[arc5.len() - 1];

    /*
        Draw path pattern:
        without morph           : M -> A1 -> C1 -> C2 -> A2
        without morph (reverse) : M -> A1 ->             A2 -> C3 -> C4
        with morph              : M -> A1 -> C1 -> C2 -> A2 -> C3 -> C4
    */
    let M = path_command::MoveTo::M(arc1_start.point.x, arc1_start.point.y);
    let A1 = draw_arcs(morphing_shape, 0, 0, &arc1[1..]);
    let A2 = draw_arcs(morphing_shape, 0, 1, &arc2[1..]);

    if morphing_shape {
        let (C1, C2, C3, C4) = match shape {
            ArcShape::Comet => {
                let C1 = path_command::CubicBezierCurve::c(
                    arc3_end.point.x,
                    arc3_end.point.y,
                    arc5_end.point.x,
                    arc5_end.point.y,
                    arc5_end.point.x,
                    arc5_end.point.y,
                );

                let C2 = path_command::CubicBezierCurve::c(
                    arc5_end.point.x,
                    arc5_end.point.y,
                    arc4_start.point.x,
                    arc4_start.point.y,
                    arc2_start.point.x,
                    arc2_start.point.y,
                );

                let C3 = path_command::CubicBezierCurve::c(
                    arc2_end.point.x,
                    arc2_end.point.y,
                    arc2_end.point.x,
                    arc2_end.point.y,
                    arc2_end.point.x,
                    arc2_end.point.y,
                );

                let C4 = path_command::CubicBezierCurve::c(
                    arc2_end.point.x,
                    arc2_end.point.y,
                    arc1_start.point.x,
                    arc1_start.point.y,
                    arc1_start.point.x,
                    arc1_start.point.y,
                );

                (C1, C2, C3, C4)
            }
            ArcShape::ReverseComet => {
                let C1 = path_command::CubicBezierCurve::c(
                    arc1_end.point.x,
                    arc1_end.point.y,
                    arc1_end.point.x,
                    arc1_end.point.y,
                    arc1_end.point.x,
                    arc1_end.point.y,
                );

                let C2 = path_command::CubicBezierCurve::c(
                    arc1_end.point.x,
                    arc1_end.point.y,
                    arc2_start.point.x,
                    arc2_start.point.y,
                    arc2_start.point.x,
                    arc2_start.point.y,
                );

                let C3 = path_command::CubicBezierCurve::c(
                    arc4_end.point.x,
                    arc4_end.point.y,
                    arc5_start.point.x,
                    arc5_start.point.y,
                    arc5_start.point.x,
                    arc5_start.point.y,
                );

                let C4 = path_command::CubicBezierCurve::c(
                    arc5_start.point.x,
                    arc5_start.point.y,
                    arc3_start.point.x,
                    arc3_start.point.y,
                    arc1_start.point.x,
                    arc1_start.point.y,
                );

                (C1, C2, C3, C4)
            }
            _ => panic!("Shape not allowed"),
        };

        format!("{M} {A1} {C1} {C2} {A2} {C3} {C4}")
    } else {
        match shape {
            ArcShape::Comet => {
                let C1 = path_command::CubicBezierCurve::c(
                    arc3_end.point.x,
                    arc3_end.point.y,
                    arc5_end.point.x,
                    arc5_end.point.y,
                    arc5_end.point.x,
                    arc5_end.point.y,
                );

                let C2 = path_command::CubicBezierCurve::c(
                    arc5_end.point.x,
                    arc5_end.point.y,
                    arc4_start.point.x,
                    arc4_start.point.y,
                    arc2_start.point.x,
                    arc2_start.point.y,
                );

                format!("{M} {A1} {C1} {C2} {A2}")
            }
            ArcShape::ReverseComet => {
                let C3 = path_command::CubicBezierCurve::c(
                    arc4_end.point.x,
                    arc4_end.point.y,
                    arc5_start.point.x,
                    arc5_start.point.y,
                    arc5_start.point.x,
                    arc5_start.point.y,
                );

                let C4 = path_command::CubicBezierCurve::c(
                    arc5_start.point.x,
                    arc5_start.point.y,
                    arc3_start.point.x,
                    arc3_start.point.y,
                    arc1_start.point.x,
                    arc1_start.point.y,
                );

                format!("{M} {A1} {A2} {C3} {C4}")
            }
            _ => panic!("Shape not allowed"),
        }
    }
}
